File: java/src/main/java/org/revenj/serialization/RevenjSerialization.java
package org.revenj.serialization;

import org.revenj.patterns.Bytes;
import org.revenj.patterns.Serialization;
import org.revenj.patterns.ServiceLocator;
import org.revenj.patterns.WireSerialization;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.lang.reflect.Type;
import java.util.Optional;

public final class RevenjSerialization implements WireSerialization {
	private final JacksonSerialization json;
	private final PassThroughSerialization passThrough;

	public RevenjSerialization(ServiceLocator locator) {
		this.json = new JacksonSerialization(locator);
		this.passThrough = new PassThroughSerialization();
	}

	@Override
	public Bytes serialize(Object value, String contentType) {
		try {
			//if (contentType == null || "application/json".equals(contentType)) {
			byte[] content = json.serializeAsBytes(value);
			return new Bytes(content, content.length);
			//}
		} catch (IOException e) {
			throw new RuntimeException(e);
		}
	}

	@Override
	public void serialize(Object value, OutputStream stream, String contentType) throws IOException {
		json.serializeTo(value, stream);
	}

	@Override
	public Object deserialize(Type type, Bytes data, String accept) throws IOException {
		return json.deserialize(type, data.content, data.length);
	}

	@Override
	public Object deserialize(Type type, InputStream stream, String accept) throws IOException {
		//if (contentType == null || "application/json".equals(contentType)) {
		return json.deserialize(type, stream);
		//}
	}

	@Override
	public <TFormat> Optional<Serialization<TFormat>> find(Class<TFormat> format) {
		if (Object.class.equals(format)) {
			return Optional.of((Serialization<TFormat>) passThrough);
		} else if (String.class.equals(format)) {
			return Optional.of((Serialization<TFormat>) json);
		}
		return Optional.empty();
	}
}

File: scala/revenj-akka/src/main/scala/net/revenj/server/RevenjSerialization.scala
package net.revenj.server

import java.io.{InputStream, OutputStream}
import java.lang.reflect.Type
import java.nio.charset.StandardCharsets

import net.revenj.Utils
import net.revenj.serialization.{DslJsonSerialization, Serialization}

import scala.util.{Failure, Try}
import scala.reflect.runtime.universe._

private[revenj] class RevenjSerialization(
  dslJson: DslJsonSerialization,
  loader: ClassLoader
) extends WireSerialization {
  private val passThrough = new PassThroughSerialization()
  private val mirror = runtimeMirror(loader)

  private val NULL = "null".getBytes(StandardCharsets.UTF_8)
  override def serialize(value: Any, stream: OutputStream, contentType: String, manifest: Type): Try[String] = {
    require(contentType == "application/json", "Only application/json content type is supported")
    if (value != null) {
      dslJson.serializeRuntime(value, stream, manifest).map(_ => "application/json")
    } else {
      Try(stream.write(NULL)).map(_ => "application/json")
    }
  }

  override def deserialize[T](manifest: Type, content: Array[Byte], length: Int, contentType: String): Try[T] = {
    require(contentType == "application/json", "Only application/json content type is supported")
    dslJson.deserializeRuntime[T](content, length, manifest)
  }

  override def deserialize[T](manifest: Type, stream: InputStream, contentType: String): Try[T] = {
    require(contentType == "application/json", "Only application/json content type is supported")
    dslJson.deserializeRuntime[T](stream, manifest)
  }

  override def deserialize[T: TypeTag](content: Array[Byte], length: Int, contentType: String): Try[T] = {
    require(contentType == "application/json", "Only application/json content type is supported")
    Utils.findType(mirror.typeOf[T], mirror) match {
      case Some(tpe) => deserialize[T](tpe, content, length, contentType)
      case _ => Failure(new IllegalArgumentException(s"Unable to find Java type for: ${mirror.typeOf[T]}"))
    }
  }

  override def deserialize[T: TypeTag](stream: InputStream, contentType: String): Try[T] = {
    require(contentType == "application/json", "Only application/json content type is supported")
    Utils.findType(mirror.typeOf[T], mirror) match {
      case Some(tpe) => deserialize[T](tpe, stream, contentType)
      case _ => Failure(new IllegalArgumentException(s"Unable to find Java type for: ${mirror.typeOf[T]}"))
    }
  }

  override def find[TFormat: TypeTag](): Option[Serialization[TFormat]] = {
    val format = mirror.typeOf[TFormat]
    if (typeOf[Any] == format || typeOf[AnyRef] == format) Some(passThrough.asInstanceOf[Serialization[TFormat]])
    else if (typeOf[String] == format) Some(dslJson.asInstanceOf[Serialization[TFormat]])
    else None
  }
}
